docs/src/main/asciidoc/config-extending-support.adoc

```diff
@@ -0,0 +1,338 @@
+= Extending Configuration Support
+
+include::./attributes.adoc[]
+
+:numbered:
+:sectnums:
+:sectnumlevels: 4
+:toc:
+
+[[custom-config-source]]
+== Custom Config Sources
+
+It is possible to create a Custom Config Source as specified in
+link:https://github.com/eclipse/microprofile-config/blob/master/spec/src/main/asciidoc/configsources.asciidoc#custom-configsources-via-configsourceprovider[MicroProfile Config Source].
+
+With a Custom Config Source it is possible to read additional configuration values and add them to the Config
+instance in a defined ordinal. This allows overriding values from other sources or falling back to other values.
+
+image::config-sources.png[]
+
+A custom Config Source requires an implementation of `org.eclipse.microprofile.config.spi.ConfigSource` or
+`org.eclipse.microprofile.config.spi.ConfigSourceProvider`. Each implementation requires registration via
+the https://docs.oracle.com/javase/8/docs/api/java/util/ServiceLoader.html[ServiceLoader] mechanism, either in
+`META-INF/services/org.eclipse.microprofile.config.spi.ConfigSource` or
+`META-INF/services/org.eclipse.microprofile.config.spi.ConfigSourceProvider`.
+
+=== Example
+
+Consider a simple in-memory ConfigSource:
+
+.org.acme.config.InMemoryConfigSource
+[source,java]
+----
+package org.acme.config;
+
+import org.eclipse.microprofile.config.spi.ConfigSource;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+public class InMemoryConfigSource implements ConfigSource {
+    private static final Map<String, String> configuration = new HashMap<>();
+
+    static {
+        configuration.put("my.prop", "1234");
+    }
+
+    @Override
+    public int getOrdinal() {
+        return 275;
+    }
+
+    @Override
+    public Set<String> getPropertyNames() {
+        return configuration.keySet();
+    }
+
+    @Override
+    public String getValue(final String propertyName) {
+        return configuration.get(propertyName);
+    }
+
+    @Override
+    public String getName() {
+        return InMemoryConfigSource.class.getSimpleName();
+    }
+}
+----
+
+And registration in:
+
+.META-INF/services/org.eclipse.microprofile.config.spi.ConfigSource
+[source,properties]
+----
+org.acme.config.InMemoryConfigSource
+----
+
+The `InMemoryConfigSource` will be ordered between the `.env` source, and the `application.properties` source due to
+the `275` ordinal:
+
+
+|===
+|ConfigSource |Ordinal
+
+|System Properties
+|400
+
+|Environment Variables from System
+|300
+
+|Environment Variables from `.env` file
+|295
+
+|InMemoryConfigSource
+|275
+
+|`application.properties` from `/config`
+|260
+
+|`application.properties` from application
+|250
+
+|`microprofile-config.properties` from application
+|100
+|===
+
+In this case, `my.prop` from `InMemoryConfigSource` will only be used if the Config Engine is unable to find a value
+in `System Properties`, `Environment Variables from System` and `Environment Variables from .env file` in this order.
```

Comment:
Why are these phrases monospaced? They don't look like code fragments to me. Should they be links?

Reply:
Yes. Let me add them.